clients/src/main/java/org/apache/kafka/common/protocol/Writable.java

```diff
@@ -38,6 +39,17 @@ default void writeNullableBytes(byte[] arr) {
         }
     }
 
+    /**
+     * Write a nullable {@link ByteBuffer} delimited by a four-byte length prefix.
+     */
+    default void writeNullableByteBuffer(ByteBuffer buf) {
+        if (buf == null) {
+            writeInt(-1);
+        } else {
+            writeBytes(buf.array());
```

Comment:
Is it safe to always assume the byte buffer is backed by the array exactly? Theoretically the same array can be used for multiple byte buffer. Maybe we should just write from offset to limit?

Reply:
> Theoretically the same array can be used for multiple byte buffer.

We can have different implementation of `ByteBuffer`, also(`DirectByteBuffer`, for example).
I reworked this method and moved it to the `Writable` implementation.